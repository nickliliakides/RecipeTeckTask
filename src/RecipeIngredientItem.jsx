import React from 'react';

const RecipeIngredientItem = ({ item }) => {
  return (
    <li className='recipe__item'>
      <svg className='recipe__icon'>
        <use href='img/icons.svg#icon-check'></use>
      </svg>
      <div className='recipe__count'>{item.count}</div>
      <div className='recipe__ingredient'>{item.ingredient}</div>
    </li>
  );
};

export default RecipeIngredientItem;
