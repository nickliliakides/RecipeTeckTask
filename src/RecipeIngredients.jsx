import React, { useEffect, useState } from 'react';
import RecipeIngredientItem from './RecipeIngredientItem';
const Fraction = require('fractional').Fraction;

const RecipeIngredients = ({ recipe, loading }) => {
  const [ingredients, setIngredients] = useState([]);
  const [currentServings, setCurrentServings] = useState();

  useEffect(() => {
    if(recipe) {
      if(!currentServings) setCurrentServings(recipe.servings)
      /* Here is handled all the logic to slplit the ingredient in two strings(quantity and the ingredient itself). If there is a fraction need to be checked as well, and myltiplied with the factor(default servings / current servings) to do the calculation everytime the user increases or decreases the servings quantity  */
      const ingredientArray = recipe.ingredients.map((ing) => {
        const ingredientsArray = ing.split(' ');
        let count = ingredientsArray[0];
        if (ingredientsArray[1].match(/^\d/)) {
          count = `${count} ${ingredientsArray[1]}`;
          ingredientsArray.splice(0, 2);
        } else {
          ingredientsArray.splice(0, 1);
        }
        if (currentServings !== recipe.servings) {
          if (count.includes('/')) {
            if (count.includes(' ')) {
              const newCountArray = count.split(' ');
              const fractionArray = newCountArray[1].split('/');
              const countNum = new Fraction(newCountArray[0]).add(
                new Fraction(fractionArray[0], fractionArray[1])
              );
  
              count = countNum
                .multiply(new Fraction(currentServings, recipe.servings))
                .toString();
            } else {
              const fractionArray = count.split('/');
              count = new Fraction(fractionArray[0], fractionArray[1])
                .multiply(new Fraction(currentServings, recipe.servings))
                .toString();
            }
          } else {
            count = new Fraction(count)
              .multiply(new Fraction(currentServings, recipe.servings))
              .toString();
          }
        }
        const ingredient = ingredientsArray.join(' ');
  
        return { count, ingredient };
      });

      setIngredients(ingredientArray);
    }
   
  }, [recipe, currentServings]);

  return recipe && !loading ? (
    <div className='shopping'>
      <div className='ingredients'>
        <h2 className='heading-2'>Ingredients</h2>
        <div className='recipe__info'>
          <svg className='recipe__info-icon'>
            <use href='img/icons.svg#icon-man'></use>
          </svg>
          <span className='recipe__info-data recipe__info-data--people'>
            {currentServings}
          </span>
          <span className='recipe__info-text'> servings</span>
          <div className='recipe__info-buttons'>
            <button
              disabled={currentServings === 1}
              className='btn-tiny btn-decrease'
              onClick={() =>
                setCurrentServings((currentServ) => {
                  return currentServ - 1;
                })
              }
            >
              <svg>
                <use href='img/icons.svg#icon-circle-with-minus'></use>
              </svg>
            </button>
            <button
              className='btn-tiny btn-increase'
              onClick={() =>{
                if(currentServings < 20) {
                  setCurrentServings((currentServ) => {
                    return currentServ + 1;
                  })
                } else {
                  alert('20 people are the greater number to calculate the ingredients')
                }
              }

              }
            >
              <svg>
                <use href='img/icons.svg#icon-circle-with-plus'></use>
              </svg>
            </button>
          </div>
        </div>

        <ul className='recipe__ingredient-list'>
          {ingredients.map((ing, i) => (
            <RecipeIngredientItem key={i} item={ing} />
          ))}
        </ul>
      </div>

      <div className='categories'>
        <h2 className='heading-2 cat'>Categories</h2>
        <ul className='recipe__tags__list'>
          {recipe.tags.map((tag, i) => (
            <li key={i}>
              <p>{tag}</p>
            </li>
          ))}
        </ul>
      </div>
    </div>
  ) : <div className='loading'>Loading ...</div>
};

export default RecipeIngredients;
