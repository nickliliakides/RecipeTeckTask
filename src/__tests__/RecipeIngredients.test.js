import React from 'react';
import { shallow } from 'enzyme';
import RecipeIngredients from '../RecipeIngredients';
import recipes from '../mockRecipes';
import RecipeIngredientItem from '../RecipeIngredientItem';

let wrapper;

const defaultProps = {
  recipe: recipes[0],
  loading: false,
};

describe('<RecipeIngredients /> component', () => {
  beforeEach(() => {
    wrapper = shallow(<RecipeIngredients {...defaultProps} />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('should matches its snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the right amount of RecipeIngredientItem components', () => {
    setTimeout(() => {
      expect(wrapper.find(RecipeIngredientItem).length).toEqual(
        defaultProps.recipe.ingredients.length
      );
    }, 500);
  });
});
