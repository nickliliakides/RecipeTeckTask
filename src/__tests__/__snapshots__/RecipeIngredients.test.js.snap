// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecipeIngredients /> component should matches its snapshot 1`] = `
<div
  className="shopping"
>
  <div
    className="ingredients"
  >
    <h2
      className="heading-2"
    >
      Ingredients
    </h2>
    <div
      className="recipe__info"
    >
      <svg
        className="recipe__info-icon"
      >
        <use
          href="img/icons.svg#icon-man"
        />
      </svg>
      <span
        className="recipe__info-data recipe__info-data--people"
      />
      <span
        className="recipe__info-text"
      >
         servings
      </span>
      <div
        className="recipe__info-buttons"
      >
        <button
          className="btn-tiny btn-decrease"
          disabled={false}
          onClick={[Function]}
        >
          <svg>
            <use
              href="img/icons.svg#icon-circle-with-minus"
            />
          </svg>
        </button>
        <button
          className="btn-tiny btn-increase"
          onClick={[Function]}
        >
          <svg>
            <use
              href="img/icons.svg#icon-circle-with-plus"
            />
          </svg>
        </button>
      </div>
    </div>
    <ul
      className="recipe__ingredient-list"
    />
  </div>
  <div
    className="categories"
  >
    <h2
      className="heading-2 cat"
    >
      Categories
    </h2>
    <ul
      className="recipe__tags__list"
    >
      <li
        key="0"
      >
        <p>
          main dish
        </p>
      </li>
      <li
        key="1"
      >
        <p>
          vegetarian
        </p>
      </li>
      <li
        key="2"
      >
        <p>
          indian
        </p>
      </li>
    </ul>
  </div>
</div>
`;
